### PR TITLE
Groth16 snarkjs proof example test

### PR DESCRIPTION
**Pull request by mstraka100**
## Description
Adds a simple example test case for verifying a groth16 proof output by the snarkjs toolchain.

## How Has This Been Tested?
The test passes

## Key Areas to Review
The new test

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [x] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [x] Other (move example)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by zjma — APPROVED** [review]

File: aptos-move/move-examples/groth16_example/sources/groth16.move
Comment (alinush) [inline aptos-move/move-examples/groth16_example/sources/groth16.move:1]: Sweet! I will take a closer look once I'm at a computer.

The proof verifies, I assume?

And can you add the JSONs in the comments please?

File: aptos-move/move-examples/groth16_example/sources/groth16.move [hunk 1]
```diff
@@ -82,6 +82,202 @@ module groth16_example::groth16 {
...
+        let public_3 = std::option::extract(&mut deserialize<bn254_algebra::Fr, bn254_algebra::FormatFrLsb>(&public_3_bytes));
+        let public_4 = std::option::extract(&mut deserialize<bn254_algebra::Fr, bn254_algebra::FormatFrLsb>(&public_4_bytes));
+        let public_5 = std::option::extract(&mut deserialize<bn254_algebra::Fr, bn254_algebra::FormatFrLsb>(&public_5_bytes));
+        let public_6 = std::option::extract(&mut deserialize<bn254_algebra::Fr, bn254_algebra::FormatFrLsb>(&public_6_bytes));
+        let public_7 = std::option::extract(&mut deserialize<bn254_algebra::Fr, bn254_algebra::FormatFrLsb>(&public_7_bytes));
+        let public_8 = std::option::extract(&mut deserialize<bn254_algebra::Fr, bn254_algebra::FormatFrLsb>(&public_8_bytes));
+
+        let public_inputs: vector<Element<bn254_algebra::Fr>> = vector[
+            public_1, public_2, public_3, public_4, public_5, public_6, public_7, public_8
+        ];
```
[the remark this answers is not in the document]
Comment (mstraka100) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:267]: The arkworks Groth16 implementation has 1 extra element in the gamma_abc vector. This is expected

File: aptos-move/move-examples/groth16_example/sources/groth16.move
Comment (mstraka100) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:1]: Yes the proof verifies. I think adding the JSON to the comments would just waste space, it should be pretty clear what the decimal constants in the code refer to.

Comment (mstraka100) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:1]: I just pushed an extra comment to clarify the only element from the VK JSON that was renamed. Everything else matches the naming convention in the snarkjs proof and vk JSON files

Comment (alinush) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:1]: Please add the JSON in the comments. Let space be wasted, in the interest of clarity.

Comment (alinush) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:1]: (You can also create a separate `groth16_circom.move` file for this, if you are worried about space?)

File: aptos-move/move-examples/groth16_example/sources/groth16.move [hunk 2]
```diff
@@ -82,6 +82,204 @@ module groth16_example::groth16 {
...
+    // Documentation for the serialization formats used can be found in `aptos-core/aptos-move/framework/aptos-stdlib/sources/cryptography/X.move`,
+    // where X is bn254_algebra for BN254 and bls12381_algebra for BLS12_381
+    // The names are preserved from the snarkjs proof and verifier key JSON file format, with the
+    // exception of "IC", which has been renamed to `vk_gamma_abc`
+    #[test(fx = @std)]
+    fun test_verify_circom_proof(fx: signer) {
+        enable_cryptography_algebra_natives(&fx);
+        let a_x = 9291141442484249183824149917322150275993152355313319552386216014158050680949u256;
+        let a_y = 4751084799539532208179359846086616641767957505361605807745261011239799367574u256;
+        let a_bytes = bcs::to_bytes<u256>(&a_x);
```
Comment (alinush) [inline aptos-move/move-examples/groth16_example/sources/groth16.move:103]: This should be `a_x_bytes`.

Comment (mstraka100) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:103]: It's mutated two lines down to be `a_bytes` - I felt this was clearer than calling it `a_x_bytes`, then mutating it to be the full bytes of `a`

File: aptos-move/move-examples/groth16_example/sources/groth16.move
Comment (mstraka100) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:1]: What about pushing the JSON files as separate files, then referencing them in the comments? I.e. `vk_example.json` and `proof_example.json`. They're long enough (~150 lines total) that copying them in the comments would make the code less readable, IMO

Comment (alinush) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:1]: Sure!

Comment (mstraka100) [reply aptos-move/move-examples/groth16_example/sources/groth16.move:1]: Jsut added them

**Review by alinush — APPROVED** [review]

**Pull request merged**